test/test_distributed.py

```diff
@@ -331,6 +334,58 @@ def test_get_rank_size_full_group(self):
         self.assertEqual(dist.get_world_size(group_id), dist.get_world_size())
         self.assertEqual(dist.get_rank(group_id), dist.get_rank())
 
+    def _test_barrier_timeout(self, group_id, timeout):
+        local_rank = dist.get_rank(group_id)
+
+        # Only execute barrier on rank == 0, causing it to timeout
+        if local_rank == 0:
+            expected_time = time.time() + timeout.total_seconds()
+            with self.assertRaisesRegex(RuntimeError, " (Timed out|closed) "):
+                dist.barrier(group_id)
+            self.assertGreaterEqual(time.time(), expected_time)
+        else:
+            time.sleep(timeout.total_seconds())
+
+    @unittest.skipIf(BACKEND != "gloo", "Only gloo backend supports timeouts")
+    @unittest.skipIf(
+        not INIT_METHOD.startswith("file://"),
+        "Requires file:// initialization method. " +
+        "Both tcp:// and env:// rely on the TCP store for which "
+        "reinitialization has proven racy."
+    )
+    def test_barrier_timeout_global(self):
+        dist.destroy_process_group()
+
+        # Explicitly pass world size to the barrier because we've
+        # just destroyed any state in torch.distributed.
+        self._barrier(wait_for=int(WORLD_SIZE))
+
+        # Reinitialize global process group
+        timeout = timedelta(seconds=0.2)
+        dist.init_process_group(
+            init_method=INIT_METHOD,
+            backend=BACKEND,
+            world_size=int(WORLD_SIZE),
+            rank=self.rank,
+            timeout=timeout,
+        )
+        self._test_barrier_timeout(dist.group.WORLD, timeout)
+
+    @skip_if_small_worldsize
```

Comment:
I don't think this will run in CI since the small_worldsize,  can you run this locally to ensure that it passes, by setting the WORLD_SIZE manually to 3 or larger in run_test.py